core-container/entry.sh

```diff
@@ -16,88 +16,487 @@
-    if kubectl create secret generic kubestellar --from-file="${PWD}/.kcp/admin.kubeconfig" --from-file="${PWD}/.kcp/external.kubeconfig" 2> /dev/null; then
-        echo " Created secret in the current namespace."
-    else
-        echo " Not running in a cluster!"
+
+# function ensure_espw() {
+#     echo "Ensuring ESPW..."
+
+#     kubectl ws root &> /dev/null
+
+#     if kubectl get Workspace "$ESPW_NAME" &> /dev/null; then
+#         echo "ESPW workspace already exists -- using it:"
+#         kubectl ws "$ESPW_NAME"
+#     else
+#         kubectl ws create "$ESPW_NAME" --enter
+#     fi
+
+#     kubectl apply -f "/home/kubestellar/config/exports"
+
+#     echo "Finished populating the ESPW with KubeStellar APIExports."
+# }
+
+
+function run_kcp() {
+    echo "--< Starting kcp >--"
+
+    echo Attempting to delete kubestellar secret...
+    (
+        KUBECONFIG=
+        ! kubectl delete secret kubestellar
+    )
+    echo "EXTERNAL_HOSTNAME=${EXTERNAL_HOSTNAME}"
+
+    # Check EXTERNAL_HOSTNAME
+    if [ -z "$EXTERNAL_HOSTNAME" ]; then
+        echo "Trying to guess the DNS from route/ingress...."
+        export EXTERNAL_HOSTNAME=$(guess_kcp_dns)
     fi
-else
-    if kubectl create secret generic kubestellar --from-file="${PWD}/.kcp/admin.kubeconfig" 2> /dev/null; then
-        echo " Created secret in the current namespace."
+    echo "EXTERNAL_HOSTNAME=${EXTERNAL_HOSTNAME}"
+
+    # Create the certificates
+    if [ -n "$EXTERNAL_HOSTNAME" ]; then
+        echo "Creating the TSL certificates"
+        # mkdir -p .kcp
+        cd .kcp
+        eval pieces_external=($(kubestellar-ensure-kcp-server-creds ${EXTERNAL_HOSTNAME}))
+        eval pieces_cluster=($(kubestellar-ensure-kcp-server-creds ${KUBESTELLAR_SERVICE})) #############
+        cd ..
+        echo "TLS certificates for server ${EXTERNAL_HOSTNAME}:"
+        echo ${pieces_external[0]}
+        echo ${pieces_external[1]}
+        echo ${pieces_external[2]}
+        echo "TLS certificates for server ${KUBESTELLAR_SERVICE}:"
+        echo ${pieces_cluster[0]}
+        echo ${pieces_cluster[1]}
+        echo ${pieces_cluster[2]}
+    fi
+
+    # Running kcp
+    if [ -n "$EXTERNAL_HOSTNAME" ]; then
+        echo "Removing existing apiserver keys... "
+        ! rm /home/kubestellar/.kcp/apiserver.* &> /dev/null # required to fix the restart
+        echo -n "Running kcp with TLS keys... "
+        kcp start --tls-sni-cert-key ${pieces_external[1]},${pieces_external[2]} --tls-sni-cert-key ${pieces_cluster[1]},${pieces_cluster[2]} & # &> kcp.log &
     else
-        echo " Not running in a cluster!"
+        echo -n "Running kcp without TLS keys... "
+        kcp start &
+    fi
+    echo Started.
+
+    # Waiting to be ready
+    echo "Waiting for ${KUBECONFIG}..."
+    while [ ! -f "${KUBECONFIG}" ]; do sleep 5; done
+    echo 'Waiting for "root:compute" workspace...'
+    until [ "$(kubectl ws root:compute 2> /dev/null)" != "" ]; do sleep 5;done
```

Comment:
I do not like the fact that we have multiple blocks of code to do this waiting, and they do it differently.
Not a blocker, but it should be cleaned up soon.